aten/src/ATen/core/TensorAccessor.h

```diff
@@ -5,45 +5,117 @@
 
 namespace at {
 
+template <typename T>
+struct DefaultPtrTraits {
+  typedef T* PtrType;
+};
+
+#ifdef __CUDACC__
+template <typename T>
+struct RestrictPtrTraits {
+  typedef T* __restrict__ PtrType;
+};
+#endif
+
+#ifndef AT_HOSTDEVICE
+#ifdef __CUDACC__
+#define AT_HOSTDEVICE __host__ __device__
+#else
+#define AT_HOSTDEVICE
+#endif
+#endif
 
-template<typename T, size_t N>
+template<typename T, size_t N, template <typename U> class PtrTraits = DefaultPtrTraits>
```

Comment:
It's a good idea to state what the new design constraints of TensorAccessorBase are (it works transparently on CUDA) to help future refactorers. Basically the stuff we discussed in the PR.